outputparser/defined.go

```diff
@@ -0,0 +1,148 @@
+package outputparser
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+
+	"github.com/tmc/langchaingo/llms"
+	"github.com/tmc/langchaingo/schema"
+)
+
+// Defined parses JSON output from an LLM into Go structs. A properly tagged
+// struct can generate TypeScript interfaces to help LLMs generate responses
+// that follow the desired JSON structure.
+type Defined[T any] struct {
+	schema string
+}
+
+func NewDefined[T any](source T) (Defined[T], error) {
+	var empty Defined[T]
+
+	sourceType := reflect.TypeOf(source)
+	if k := sourceType.Kind(); k != reflect.Struct {
+		return empty, fmt.Errorf("expected a struct; got %s", k)
+	}
+	numFields := sourceType.NumField()
+	if numFields == 0 {
+		return empty, errors.New("schema source has no fields")
+	}
+
+	numTaggedFields := 0
+	for i := 0; i < numFields; i++ {
+		if _, ok := sourceType.Field(i).Tag.Lookup("describe"); ok {
+			numTaggedFields++
+		}
+	}
+	if numTaggedFields == 0 {
+		return empty, errors.New("requires at least 1 field with the struct tag 'describe'")
```

Comment:
I feel as though we should still operate with zero tagged fields.

Reply:
That seems reasonable. I've updated the code, tests, and doc comments accordingly.